File: packages/quran-engine-swift/Sources/QuranEngine/Semantic.swift
import Foundation

/// Meaning-based ("AI") search: find the ayahs about a topic whether or not they use its words.
///
/// WHY WORD VECTORS AND MaxSim, NOT A SENTENCE EMBEDDING - measured, not assumed. Scoring an ayah
/// by the cosine between a SENTENCE embedding of the query and one of the ayah ranks this corpus
/// close to randomly: translated scripture is dense, and one vector for a whole verse washes out
/// the single idea the query is asking about. Scoring word by word fixes it - embed every word, and
/// score a text as the MEAN over the query's words of the BEST matching word in the text. On real
/// verses that separates related (0.42-0.70) from unrelated (0.27-0.41) cleanly, and it degrades
/// gracefully: a query word the model has never seen contributes nothing instead of poisoning the
/// vector.
///
/// THE EMBEDDER IS YOURS. This engine ships no model - word vectors are tens of megabytes and every
/// platform already has one worth using. On Apple platforms that is `NLEmbedding.wordEmbedding(for:)`:
///
/// ```swift
/// import NaturalLanguage
/// let embedding = NLEmbedding.wordEmbedding(for: .english)
/// let semantic = Semantic { word in embedding?.vector(for: word)?.map(Float.init) }
/// ```
///
/// Build the index once and keep it: over the 6,236 translations that is seconds and ~10-25 MB,
/// which is why any serious consumer persists it rather than rebuilding per launch.
///
/// See docs/14-ask-ai.md.
public struct SemanticHit: Sendable, Equatable {
    public let id: String
    /// Mean best-match cosine over the query's words, 0...1.
    public let score: Float
}

public final class Semantic {
    public typealias Embedder = (String) -> [Float]?

    private let embed: Embedder
    private let minWordLength: Int
    private var vectors: [String: [Float]?] = [:]
    private var documents: [(id: String, vectors: [[Float]])] = []

    public init(minWordLength: Int = 3, embed: @escaping Embedder) {
        self.embed = embed
        self.minWordLength = minWordLength
    }

    public var size: Int { documents.count }

    /// Build (or rebuild) the index. Call once per corpus.
    @discardableResult
    public func index(_ corpus: [(id: String, text: String)]) -> Semantic {
        documents = corpus.compactMap { document in
            let vectors = vectorize(document.text)
            return vectors.isEmpty ? nil : (id: document.id, vectors: vectors)
        }
        return self
    }

    /// The documents closest in meaning to `query`, best first. `minScore` is a floor on "actually
    /// related" - 0.42 is a sensible start on English translations, but calibrate it against YOUR
    /// embedder.
    public func search(_ query: String, limit: Int = 10, minScore: Float = 0) -> [SemanticHit] {
        let queryVectors = vectorize(query)
        guard !queryVectors.isEmpty else { return [] }
        var hits: [SemanticHit] = []
        for document in documents {
            var total: Float = 0
            for q in queryVectors {
                var best: Float = -1
                for w in document.vectors {
                    let score = Semantic.cosine(q, w)
                    if score > best { best = score }
                }
                total += best
            }
            let score = total / Float(queryVectors.count)
            if score >= minScore { hits.append(SemanticHit(id: document.id, score: score)) }
        }
        hits.sort { $0.score != $1.score ? $0.score > $1.score : $0.id < $1.id }
        return Array(hits.prefix(limit))
    }

    public func clear() {
        documents = []
        vectors = [:]
    }

    /// Cosine similarity of two ALREADY NORMALIZED vectors, i.e. their dot product.
    public static func cosine(_ a: [Float], _ b: [Float]) -> Float {
        var sum: Float = 0
        for i in 0..<min(a.count, b.count) { sum += a[i] * b[i] }
        return sum
    }

    private func vectorize(_ text: String) -> [[Float]] {
        var out: [[Float]] = []
        var seen = Set<String>()
        for raw in text.lowercased().split(whereSeparator: { !$0.isLetter && !$0.isNumber }) {
            let word = String(raw)
            guard word.count >= minWordLength, seen.insert(word).inserted else { continue }
            if let vector = vector(for: word) { out.append(vector) }
        }
        return out
    }

    private func vector(for word: String) -> [Float]? {
        if let cached = vectors[word] { return cached }
        let raw = embed(word)
        // Normalized once here, so scoring is a dot product rather than three passes per pair.
        let normalized = (raw?.isEmpty == false) ? Semantic.normalize(raw!) : nil
        vectors[word] = normalized
        return normalized
    }

    private static func normalize(_ raw: [Float]) -> [Float] {
        let magnitude = sqrt(raw.reduce(0) { $0 + $1 * $1 })
        guard magnitude > 0 else { return Array(repeating: 0, count: raw.count) }
        return raw.map { $0 / magnitude }
    }
}

File: packages/quran-engine-js/src/semantic.js
// @ts-check
/**
 * Meaning-based ("AI") search: find the ayahs about a topic whether or not they use its words.
 * "Patience in hardship" should reach the verses about sabr even when neither word appears.
 *
 * WHY WORD VECTORS AND MaxSim, NOT A SENTENCE EMBEDDING
 * -----------------------------------------------------
 * Measured, not assumed. Scoring an ayah by the cosine between a SENTENCE embedding of the query
 * and one of the ayah ranks this corpus close to randomly - translated scripture is dense, and one
 * vector for a whole verse washes out the one idea the query is asking about. Scoring word by word
 * fixes it: embed every word, and score a text as the MEAN over the query's words of the BEST
 * matching word in the text. On real verses that separates related (0.42-0.70) from unrelated
 * (0.27-0.41) cleanly, and it degrades gracefully - a query word the model has never seen simply
 * contributes nothing instead of poisoning the vector.
 *
 * THE EMBEDDER IS YOURS
 * ---------------------
 * This engine ships no model: word vectors are tens of megabytes and every platform already has
 * one worth using (Apple's `NLEmbedding.wordEmbedding`, Android's ML Kit, `wink-embeddings` or a
 * GloVe file in Node, whatever your stack offers). Hand `embed` a function from a lowercased word
 * to its vector, or null when it has none, and this module does the rest. Vectors are cached per
 * word, so a repeated word costs one lookup for the whole corpus.
 *
 * Build the index once (it is O(words in corpus) embedder calls) and keep it: over the 6,236
 * English translations that is a few seconds and ~10-25 MB, which is why every serious consumer
 * persists it rather than rebuilding per launch.
 */

const WORD = /[^\p{L}\p{N}]+/u;

/**
 * @typedef {Object} SemanticHit
 * @property {string} id
 * @property {number} score  mean best-match cosine over the query's words, 0...1
 * @property {any} [meta]
 */

export class Semantic {
  /**
   * @param {Object} opts
   * @param {(word: string) => (number[]|Float32Array|null|undefined)} opts.embed
   * @param {number} [opts.minWordLength=3] words shorter than this are skipped on both sides
   */
  constructor({ embed, minWordLength = 3 }) {
    if (typeof embed !== "function") throw new TypeError("Semantic needs an embed(word) function");
    this._embed = embed;
    this._minWordLength = minWordLength;
    /** @type {Map<string, Float32Array|null>} */
    this._vectors = new Map();
    /** @type {{id:string, meta:any, vectors:Float32Array[]}[]} */
    this._documents = [];
  }

  /** How many documents are indexed. */
  get size() {
    return this._documents.length;
  }

  /**
   * Build (or rebuild) the index. Call once per corpus.
   * @param {Iterable<{id:string, text:string, meta?:any}>} documents
   */
  index(documents) {
    this._documents = [];
    for (const doc of documents) {
      const vectors = this._vectorize(doc.text);
      if (vectors.length) this._documents.push({ id: doc.id, meta: doc.meta ?? null, vectors });
    }
    return this;
  }

  /**
   * The documents closest in meaning to `query`, best first.
   * @param {string} query
   * @param {Object} [opts]
   * @param {number} [opts.limit=10]
   * @param {number} [opts.minScore=0] drop hits below this - 0.42 is a sensible "actually related"
   *        floor on English translations, but calibrate it against YOUR embedder
   * @returns {SemanticHit[]}
   */
  search(query, { limit = 10, minScore = 0 } = {}) {
    const queryVectors = this._vectorize(query);
    if (!queryVectors.length) return [];

    /** @type {SemanticHit[]} */
    const hits = [];
    for (const doc of this._documents) {
      let total = 0;
      for (const q of queryVectors) {
        let best = -1;
        for (const w of doc.vectors) {
          const score = cosine(q, w);
          if (score > best) best = score;
        }
        total += best;
      }
      const score = total / queryVectors.length;
      if (score >= minScore) hits.push({ id: doc.id, score, meta: doc.meta });
    }
    hits.sort((a, b) => b.score - a.score || (a.id < b.id ? -1 : 1));
    return hits.slice(0, limit);
  }

  /** Drop the index and the vector cache. */
  clear() {
    this._documents = [];
    this._vectors.clear();
    return this;
  }

  /** @param {string} text @returns {Float32Array[]} */
  _vectorize(text) {
    /** @type {Float32Array[]} */
    const out = [];
    const seen = new Set();
    for (const raw of text.toLowerCase().split(WORD)) {
      if (raw.length < this._minWordLength || seen.has(raw)) continue;
      seen.add(raw);
      const vector = this._vector(raw);
      if (vector) out.push(vector);
    }
    return out;
  }

  /** @param {string} word */
  _vector(word) {
    if (this._vectors.has(word)) return this._vectors.get(word) ?? null;
    const raw = this._embed(word);
    // Normalized once here, so scoring is a dot product rather than three passes per pair.
    const vector = raw && raw.length ? normalize(raw) : null;
    this._vectors.set(word, vector);
    return vector;
  }
}

/**
 * Cosine similarity of two ALREADY NORMALIZED vectors, i.e. their dot product.
 * @param {Float32Array} a @param {Float32Array} b
 */
export function cosine(a, b) {
  const n = Math.min(a.length, b.length);
  let sum = 0;
  for (let i = 0; i < n; i++) sum += a[i] * b[i];
  return sum;
}

/** @param {number[]|Float32Array} raw */
function normalize(raw) {
  let magnitude = 0;
  for (let i = 0; i < raw.length; i++) magnitude += raw[i] * raw[i];
  magnitude = Math.sqrt(magnitude);
  const out = new Float32Array(raw.length);
  if (magnitude === 0) return out;
  for (let i = 0; i < raw.length; i++) out[i] = raw[i] / magnitude;
  return out;
}
